File: backend/app/routers/recipes.py
"""Recipes router — the one-click Mission gallery (read-only).

A Recipe is a goal-oriented, mission-mode skill promoted into a browsable gallery
(see docs/ALLY-RECIPES.md). This endpoint just lists the eligible skills and their
form variables; the frontend fills the variables into ``goal_template`` and sends the
resulting sentence through the normal chat → mission-offer → Start pipeline. There is
deliberately no "run recipe" endpoint here — nothing bypasses the chat/mission engine,
so all per-step safety validation, the verification gate, budgets, and persistence keep
working exactly as already built.
"""
from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel

from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies.access import resolve_server
from app.dependencies.auth import get_current_user
from app.models.user import User
from app.services import skill_service

router = APIRouter(prefix="/api/recipes", tags=["recipes"])


class RecipeVariable(BaseModel):
    name: str
    required: bool
    default: str = ""


class RecipeOut(BaseModel):
    slug: str
    title: str
    summary: str
    icon: str
    os_family: str
    budget: int
    variables: list[RecipeVariable]
    goal_template: str


@router.get("", response_model=list[RecipeOut])
async def list_recipes(
    os: str | None = Query(default=None, description="target server os_type to gate against"),
    server_id: str | None = Query(default=None,
                                  description="gate against a specific server as well"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[RecipeOut]:
    """List goal-oriented recipes, gated against the target the customer has chosen.

    Passing a server gates on what that machine actually is, not just its OS — two
    recipes can answer "host a website" and only one of them applies to a server with a
    control panel. The same rule the chat router uses, so a customer is never offered a
    recipe that would refuse the moment it started.
    """
    panel = None
    if server_id:
        server = await resolve_server(server_id, current_user, db)
        os = os or server.os_type
        # `or ""` on purpose: a null panel on a real server means "no panel", not
        # "we do not know" — see skill_service.server_ok.
        panel = server.panel_type or ""
    return [
        RecipeOut(
            slug=s.slug,
            title=s.title,
            summary=s.summary or s.title,
            icon=s.icon,
            os_family=s.os_family,
            budget=skill_service.resolve_mission_budget(s),
            variables=[RecipeVariable(**v) for v in s.variables],
            goal_template=s.goal_template,
        )
        for s in skill_service.list_recipes(os)
        if skill_service.server_ok(s, panel)
    ]

File: frontend/src/api/recipes.ts
import { apiClient } from "./client"

export interface RecipeVariable {
  name: string
  required: boolean
  default: string
}

/** A Recipe = a goal-oriented mission skill promoted into the one-click gallery. */
export interface Recipe {
  slug: string
  title: string
  summary: string
  icon: string
  os_family: string // 'linux' | 'windows' | 'any'
  budget: number
  variables: RecipeVariable[]
  goal_template: string
}

/** List recipes, optionally OS-gated against a target server's os_type. */
/** Pass a server to get only the recipes that apply to THAT machine — two recipes can
 *  answer "host a website" and only one fits a server with a control panel. */
export async function listRecipes(
  os?: string | null, serverId?: string | null,
): Promise<Recipe[]> {
  const params: Record<string, string> = {}
  if (os) params.os = os
  if (serverId) params.server_id = serverId
  const { data } = await apiClient.get<Recipe[]>("/api/recipes", {
    params: Object.keys(params).length ? params : undefined,
  })
  return data
}

/**
 * Fill a recipe's variables into its goal_template, producing the chat message that
 * starts the mission. Runs entirely client-side (like RunPlaybookModal's access-card
 * fill) — nothing round-trips the backend until it's sent as a normal chat message.
 *
 * Values are resolved in the recipe's declared order so a later variable's default may
 * reference an earlier one (e.g. `title` defaults to `{{domain}}`). A blank optional
 * with no default is dropped from the sentence rather than left as an empty `''`.
 */
export function composeRecipeMessage(recipe: Recipe, values: Record<string, string>): string {
  const resolved: Record<string, string> = {}
  const fillRefs = (tpl: string): string =>
    tpl.replace(/\{\{(\w+)\}\}/g, (_m, k: string) => resolved[k] ?? "")
  for (const v of recipe.variables) {
    let val = (values[v.name] ?? "").trim()
    if (!val && v.default) val = fillRefs(v.default).trim()
    resolved[v.name] = val
  }
  return fillRefs(recipe.goal_template).replace(/\s+/g, " ").trim()
}

/** Required variables the user left blank — block submit until empty. */
export function missingRequired(recipe: Recipe, values: Record<string, string>): string[] {
  return recipe.variables
    .filter((v) => v.required && !(values[v.name] ?? "").trim())
    .map((v) => v.name)
}
